ssz/sedes/container.py

```diff
@@ -1,104 +1,163 @@
 from typing import (
+    IO,
     Any,
-    Dict,
-    Generator,
+    Iterable,
     Sequence,
     Tuple,
-    TypeVar,
 )
 
 from eth_utils import (
     ValidationError,
-    to_dict,
+    to_tuple,
 )
-from mypy_extensions import (
-    TypedDict,
+from eth_utils.toolz import (
+    sliding_window,
 )
 
 from ssz.exceptions import (
     DeserializationError,
+    SerializationError,
 )
 from ssz.sedes.base import (
-    BaseSedes,
     CompositeSedes,
+    TSedes,
 )
 from ssz.utils import (
-    get_duplicates,
     merkleize,
+    read_exact,
+    s_decode_offset,
 )
 
-AnyTypedDict = TypedDict("AnyTypedDict", {})
-TAnyTypedDict = TypeVar("TAnyTypedDict", bound=AnyTypedDict)
 
+@to_tuple
+def _deserialize_fixed_size_items_and_offsets(stream, field_sedes):
+    for sedes in field_sedes:
+        if sedes.is_fixed_sized:
+            field_size = sedes.get_fixed_size()
+            field_data = read_exact(field_size, stream)
+            yield (sedes.deserialize(field_data), sedes)
+        else:
+            yield (s_decode_offset(stream), sedes)
 
-class Container(CompositeSedes[TAnyTypedDict, Dict[str, Any]]):
-    def __init__(self, fields: Sequence[Tuple[str, BaseSedes[Any, Any]]]) -> None:
-        self.fields = fields
-        self.field_names = tuple(field_name for field_name, _ in self.fields)
-        self.field_sedes_objects = tuple(field_sedes for _, field_sedes in self.fields)
-        self.field_name_to_sedes = dict(self.fields)
 
-        if len(fields) == 0:
+class Container(CompositeSedes[Sequence[Any], Tuple[Any, ...]]):
+    def __init__(self, field_sedes: Sequence[TSedes]) -> None:
+        if len(field_sedes) == 0:
             raise ValidationError("Cannot define container without any fields")
-
-        duplicate_field_names = get_duplicates(self.field_names)
-        if duplicate_field_names:
-            raise ValidationError(
-                f"The following fields are duplicated {','.join(sorted(duplicate_field_names))}"
-            )
+        self.field_sedes = tuple(field_sedes)
```

Comment:
Ouch, just realize that, with this change, we can't [hack vector length with `Container.field_name_to_sedes` dict directly](https://github.com/ethereum/trinity/blob/master/eth2/beacon/tools/misc/ssz_vector.py), and the field names are not available from `to_formatted_dict()` result.

1. Is there any chance to keep the field names available in `Container`? It can be `field_names: Tuple[str, ...]` instead of `field_name_to_sedes: Dict[str, Sequence[TSedes]]` if you want to make it immutable if we can still do forced update the length in Trinity tests.
2. IMO having field names in `to_formatted_dict()` result is a requirement.

Reply:
@hwwhww Yes, this should be updated to return a dictionary with the names intact.  I'll make that change.

Reply:
Delete my previous response.

I **do not** think that `Container` should  know about field names, however, I think that `Serializable` can adjust for this.

We can updated the `to_formatted_dict` implementation for `Serializables` to return a `Mapping`, preserving the field names.  Does that solve your problem?